core/src/test/scala/org/apache/spark/deploy/DecommissionWorkerSuite.scala

```diff
@@ -0,0 +1,424 @@
+    createWorkers(2)
+    val ss = new ExternalShuffleServiceHolder(conf)
+    sc = createSparkContext(conf)
+
+    // Here we will create a 2 stage job: The first stage will have two tasks and the second stage
+    // will have one task. The two tasks in the first stage will be long and short. We decommission
+    // and kill the worker after the short task is done. Eventually the driver should get the
+    // executor lost signal for the short task executor. This should trigger regenerating
+    // the shuffle output since we cleanly decommissioned the executor, despite running with an
+    // external shuffle service.
+    try {
+      val executorIdToWorkerInfo = getExecutorToWorkerAssignments
+      val workerForTask0Decommissioned = new AtomicBoolean(false)
+      // single task job
+      val listener = new RootStageAwareListener {
+        override def handleRootTaskEnd(taskEnd: SparkListenerTaskEnd): Unit = {
+          val taskInfo = taskEnd.taskInfo
+          delayedAssert(taskInfo.index <= 1, s"Expected ${taskInfo.index} <= 1")
+          delayedAssert(taskInfo.successful, s"Task ${taskInfo.index} should be successful")
+          if (taskInfo.index == 0) {
+            if (workerForTask0Decommissioned.compareAndSet(false, true)) {
+              // Since this task hasn't been killed before, it should still be at its first attempt.
+              delayedAssert(taskInfo.attemptNumber === 0, "Should have succeeded in 1st attempt")
+              val workerInfo = executorIdToWorkerInfo(taskInfo.executorId)
+              decommissionWorkerOnMaster(workerInfo, "Kill early done map worker")
+              killWorkerAfterTimeout(workerInfo, 0)
+              logInfo(s"Killed the node ${workerInfo.hostPort} that was running the early task")
+            } else {
+              // The task should have been rerun since the worker was decommissioned just after
+              // it was finished.
+              // either the task attempt or the stage attempt number should be more than 0.
+              val attemptNumber = taskInfo.attemptNumber
+              val stageAttempt = taskEnd.stageAttemptId
+              delayedAssert(attemptNumber > 0 || stageAttempt > 0,
+                s"The task should have been rerun after decommissioning worker:" +
+                  s" ($attemptNumber, $stageAttempt)")
+            }
+          } else {
+            delayedAssert(taskInfo.attemptNumber === 0, "2nd task should succeed on 1st attempt")
+          }
+        }
+      }
+      sc.addSparkListener(listener)
+      val jobResult = sc.parallelize(1 to 2, 2).mapPartitionsWithIndex((pid, _) => {
+        val sleepTimeSeconds = if (pid == 0) 1 else 10
+        Thread.sleep(sleepTimeSeconds * 1000L)
+        List(1).iterator
+      }, preservesPartitioning = true).repartition(1).sum()
+      assert(jobResult === 2)
+      // 4 tasks: 2 from first stage, one retry due to decom, one more in the second stage.
```

Comment:
This line explains the 4 tasks quite well but the assert 2 lines below checks `>= 4` so please mention what could be the addition task